src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java

```diff
@@ -167,4 +169,123 @@ private static int doSwitch(Object target, int startIndex, Object[] labels) {
+     *               privileges of the caller.  When used with {@code invokedynamic},
+     *               this is stacked automatically by the VM.
+     * @param invocationName unused
+     * @param invocationType The invocation type of the {@code CallSite} with two parameters,
+     *                       an enum type, an {@code int}, and {@code int} as a return type.
+     * @param labels case labels - {@code String} constants and {@code Class} instances,
+     *               in any combination
+     * @return a {@code CallSite} returning the first matching element as described above
+     *
+     * @throws NullPointerException if any argument is {@code null}
+     * @throws IllegalArgumentException if any element in the labels array is null, if the
+     * invocation type is not a method type of first parameter of an enum type,
+     * second parameter of type {@code int} and with {@code int} as its return type,
+     * or if {@code labels} contains an element that is not of type {@code String} or
+     * {@code Class} of the target enum type.
+     * @jvms 4.4.6 The CONSTANT_NameAndType_info Structure
+     * @jvms 4.4.10 The CONSTANT_Dynamic_info and CONSTANT_InvokeDynamic_info Structures
+     */
+    public static CallSite enumSwitch(MethodHandles.Lookup lookup,
+                                      String invocationName,
+                                      MethodType invocationType,
+                                      Object... labels) throws NullPointerException,
+                                                               IllegalArgumentException {
+        if (invocationType.parameterCount() != 2
+            || (!invocationType.returnType().equals(int.class))
+            || invocationType.parameterType(0).isPrimitive()
+            || !invocationType.parameterType(0).isEnum()
+            || !invocationType.parameterType(1).equals(int.class))
+            throw new IllegalArgumentException("Illegal invocation type " + invocationType);
+        requireNonNull(labels);
+
+        labels = labels.clone();
+
+        Class<?> enumClass = invocationType.parameterType(0);
+        labels = Stream.of(labels).map(l -> convertEnumConstants(lookup, enumClass, l)).toArray();
+
+        MethodHandle target =
+                MethodHandles.insertArguments(DO_ENUM_SWITCH, 2, (Object) labels);
+        target = target.asType(invocationType);
+
+        return new ConstantCallSite(target);
+    }
+
+    private static <E extends Enum<E>> Object convertEnumConstants(MethodHandles.Lookup lookup, Class<?> enumClassTemplate, Object label) {
+        if (label == null) {
+            throw new IllegalArgumentException("null label found");
+        }
+        Class<?> labelClass = label.getClass();
+        if (labelClass == Class.class) {
+            if (label != enumClassTemplate) {
+                throw new IllegalArgumentException("illegal Class label: " + label);
```

Comment:
Can we refine the message to state the class label is not equal to the enum class that is the target of the switch? LIkewise in the `else` block to state the label is not of class String or the target enum class.